arangod/Aql/Optimizer/Rule/OptimizerRulePushLimitIntoIndex.cpp

```diff
@@ -0,0 +1,255 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2014-2024 ArangoDB GmbH, Cologne, Germany
+/// Copyright 2004-2014 triAGENS GmbH, Cologne, Germany
+///
+/// Licensed under the Business Source License 1.1 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     https://github.com/arangodb/arangodb/blob/devel/LICENSE
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+////////////////////////////////////////////////////////////////////////////////
+
+#include "Aql/Ast.h"
+#include "Aql/AstNode.h"
+#include "Aql/Condition.h"
+#include "Aql/Expression.h"
+#include "Aql/ExecutionNode/CalculationNode.h"
+#include "Aql/ExecutionNode/IndexNode.h"
+#include "Aql/ExecutionNode/LimitNode.h"
+#include "Aql/ExecutionNode/SortNode.h"
+#include "Aql/Optimizer.h"
+#include "Aql/OptimizerRules.h"
+#include "Aql/OptimizerUtils.h"
+#include "Aql/SortElement.h"
+#include "Aql/Variable.h"
+#include "Indexes/Index.h"
+
+#include <algorithm>
+#include <optional>
+
+using namespace arangodb;
+using namespace arangodb::aql;
+using EN = arangodb::aql::ExecutionNode;
+
+/// index is eligible only if it is a type of persistent index
+/// and if it build on multiple attributes
+bool isEligibleIndex(transaction::Methods::IndexHandle const& idx) {
+  // we only care about persistent indexes.
+  // note that "hash" and "skiplist" indexes are the same as persistent
+  // indexes under the hood, just with legacy naming.
+  if (idx->type() != Index::IndexType::TRI_IDX_TYPE_PERSISTENT_INDEX &&
+      idx->type() != Index::IndexType::TRI_IDX_TYPE_HASH_INDEX &&
+      idx->type() != Index::IndexType::TRI_IDX_TYPE_SKIPLIST_INDEX) {
+    return false;
+  }
+
+  // we only care about compound indexes
+  if (idx->fields().size() < 2) {
+    return false;
+  }
+
+  return true;
+}
+
+/// optimization can be applied if the sorting attributes being sorted match
+/// the index attributes, and they are all in the same order (ASC/DESC)
+bool isEligibleSort(auto itIndex, auto const itIndexEnd, auto const& sortFields,
+                    auto const* executionPlan, auto const* indexNode) {
+  std::optional<bool> sortAscending;
+  auto itSort = sortFields.begin();
+
+  while (itIndex != itIndexEnd && itSort != sortFields.end()) {
+    auto const* sortVar = itSort->var;
```

Comment:
We are relying here on `SortElement::var`. This is fine.
However, we are not checking if `SortElement::attributePath` is set or not. I think the behavior is different depending on whether the attributePath is set or not, and we are currently not taking this into account.